SP800-108 CTR HMAC

vcsjones opened the pull request:
This implements SP800-108 CTR HMAC as a package. In .NET 8, this is an inbox component. For down-level platforms, `Microsoft.Bcl.Cryptography` was introduced as a package for .NET Standard and .NET Framework.

Some implementation notes.

1. This at runtime decides if the platform is Windows 8+. If it is, you get the CNG implementation. If it isn't, you get the Managed implementation.
2. Both the CNG and Managed implementation support NetCoreApp and .NET Standard. This results in some backward desirable behavior. The .NET Standard implementation prefers working with `byte[]` for inputs because of `IncrementalHash`. .NET Core prefers working with spans. Because of this, we avoid the `byte[]` APIs deferring to `{ReadOnly}Span<byte>` APIs. Otherwise, we will go from `byte[]` -> span -> `byte[]`. This is true not just for public APIs that accept input, but also for places we need to do conversions, like from `string` to `byte[]`. In .NET Standard, we always rent from the pool and avoid doing work on the stack, whereas .NET Core can use the stack.
3. This class is **thread safe** for all implementations. This includes instance methods. Instances methods should either a) product the correct result or b) throw `ObjectDisposedException`. We get this for free with safe handles and the CNG implementation. For the managed implementation, the key is reference counted. When the reference count reaches zero, the key is zeroed and attempts to acquire the key afterward throw an `ObjectDisposedException`.
4. The test data draws from a few different places (explained in the tests) such as RFCs and other implementations like SymCrypt. While NIST has CAVP test vectors, those tests use 800-108 CTR HMAC in a mode that our public API does not support. All of the test vectors assume you have total control over the input into the PRF, whereas our public API forces the input to be `label || 0x00 || context`. So the NIST test vectors are unusable for tests.

Closes #65577

stephentoub commented:
Thanks.

We really need a new package/assembly for this one type?  We can't, for example, just squint and say it's related to hashing so it can go in System.IO.Hashing?

vcsjones commented:
> We really need a new package/assembly for this one type?

There is internal (from Microsoft) interest in this capability for .NET Framework, and the API proposal suggested this as a package name.

> We can't, for example, just squint and say it's related to hashing so it can go in System.IO.Hashing?

My 2c: that's... quite a bit of a squint. That package is specifically documented as "non cryptographic hashing" in its README. This API is 1. cryptographic and 2. doesn't do hashing, strictly speaking (though the PRF is HMAC in this one case).

Thinking about it more, I don't really see a good "home" for this type in an existing package, so a new package makes sense to me.

Though I will defer to y'all.

stephentoub commented:
If we really lack a good existing package/assembly for it, then we should think hard about what to name this one so we don't need to create yet another the next time we need to ship another cryptographic algorithm downlevel.

vcsjones commented:
>  we should think hard about what to name this one so we don't 

@terrajobst @bartonjs @GrabYourPitchforks do y'all have any thoughts here?

vcsjones commented:
> System.Security.Cryptography.NETStandard

I would probably avoid calling anything `NETStandard`, but I don't have a big preference for what the generic name is.

I do have a bit of a preference for using a generic name at all. One thing I don't really want is a big invitation / debate on what should be in `NETStandard` or whatever going forward. Let's say, of the sake of argument, HKDF doesn't in .NET today. Would we put it in this `Extensions` or `NETStandard` package? What about `CertificateRevocationListBuilder`?

I think there is clear evidence that SP800108 needs to be made available in downlevel targets. What's not clear to me is why everything else couldn't argue that criteria.

> I don't remember if a super-generic bucket like that was proposed and shot down, or if I'm striking gold here.

It was discussed in the API review here. https://youtu.be/bDTLx5SXusY?t=5714 (1:35:20 in).

My take away from that discussion a slight preference to avoid a "dumping ground" package and "fine with another one off package".

GrabYourPitchforks commented on src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationCng.cs:
```diff
@@ -0,0 +1,212 @@
+
+        public void DeriveBytes(ReadOnlySpan<char> label, ReadOnlySpan<char> context, Span<byte> destination)
+        {
+            using (Utf8DataEncoding labelData = new Utf8DataEncoding(label, stackalloc byte[128]))
```
Consider pulling this out as a const so that it can be updated everywhere by changing a single line. Also consider what a common label \& context length might be. A scratch buffer of 128 bytes means that any source span \>= 43 chars will hit the crypto pool.

vcsjones replied:
Labels are mostly driven by RFCs or are usually on the small side. Contexts... can really go anywhere in size. I ended up bumping it to 256 because we should have the stack space for it.

GrabYourPitchforks commented on src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/NetStandardShims.cs:
```diff
@@ -0,0 +1,34 @@
+{
+    internal static class NetStandardShims
+    {
+        internal static unsafe int GetBytes(this Encoding encoding, ReadOnlySpan<char> str, Span<byte> destination)
```
I wonder if this should be moved to a common helper file and shared across projects. System.Text.Json has similar code, as shown below. (And both this file and the JSON copy have subtle bugs when handling zero-length source \& destination, which doesn't matter _here_ but we should probably address nonetheless if we're going to get in the habit of copying this code.)

https://github.com/dotnet/runtime/blob/c6d6b19da5427ae78eab241d2af596a3320a5e5b/src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs#L247-L258

vcsjones replied:
The subtle bug being... what? When a BOM needs to get written?

GrabYourPitchforks replied:
> The subtle bug being... what? When a BOM needs to get written?

The `GetChars` method can't handle null inputs or outputs, and _in the general case_ there's no guarantee that a zero-length input produces a zero-length output. Doesn't matter for UTF-8, but I'm seeing this code duplicated in a few places and we may as well just write it once and get it correct from the beginning.

vcsjones replied:
That example is not exactly the same. That one is doing `GetChars`. We are doing `GetBytes`. So we can't really consolidate these two specific examples. I can extract out helpers if you want, but in this particular case it won't result in de-duping anything.

bartonjs commented on src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationCng.NetStandard.cs:
```diff
@@ -0,0 +1,113 @@
+
+            if (key.Length > hashAlgorithmBlockSize)
+            {
+                clearSpan = HashOneShot(hashAlgorithm, key);
```
The only difference between the `byte[]` and `ReadOnlySpan<byte>` versions is the HashOneShot method call.  Can we collapse these into one big body via something like

```
_keyHandle = OpenKeyHandle(key, hashAlgorithm);

...

private static SafeSomethingHandle OpenKeyHandle(byte[] key, HashAlgorithmName hashAlgorithmName)
{
    int hashAlgorithmBlockSize = GetHashBlockSize(hashAlgorithm.Name);

    if (key.Length > hashAlgorithmBlockSize)
    {
        ...
        return OpenKeyHandleCore(hashAlgorithm, hashedKey);
    }

    return OpenKeyHandleCore(hashAlgorithm, key.AsSpan());
}

private static SafeSomethingHandle OpenKeyHandle(ReadOnlySpan<byte> key, HashAlgorithmName hashAlgorithmName)
{
    int hashAlgorithmBlockSize = GetHashBlockSize(hashAlgorithm.Name);

    if (key.Length > hashAlgorithmBlockSize)
    {
        ...
        return OpenKeyHandleCore(hashAlgorithm, hashedKey);
    }

    return OpenKeyHandleCore(hashAlgorithm, key);
}
```

Then the Core just gets to funnel down the "don't pass null", and is common for both paths.

vcsjones replied:
I tried this originally and it doesn't really improve much. In addition to the hashing, the clearing of the hashed key makes it difficult to mix and match `byte[]`, `Span<byte>` and `ReadOnlySpan<byte>`. As I worked on refactoring and fixing it, I pretty much just ended up right where I was again.

GrabYourPitchforks commented on src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs:
```diff
@@ -0,0 +1,661 @@
+        // Windows' BCryptKeyDerivation cannot fullfill a request larger than 0x1FFFFFFF, either.
+        private const int MaxPrfOutputSize = (int)(uint.MaxValue / 8);
+
+        private readonly ISP800108HmacCounterKdfImplementation _implementation;
```
Nit: For internal use, prefer abstract base classes instead of interfaces where possible. Just makes the codegen a tiny bit cleaner.

(If you have other reasons for preferring an interface, don't fret too much on this.)

vcsjones replied:
> If you have other reasons for preferring an interface

Originally I did some cute things with static abstracts on the interface, and later remembered that did not work on downlevel platforms, so no good reason.

GrabYourPitchforks commented on src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs:
```diff
@@ -0,0 +1,661 @@
+        ///   <paramref name="hashAlgorithm"/> is not a known or supported hash algorithm.
+        /// </exception>
+        /// <exception cref="EncoderFallbackException">
+        ///   <paramref name="label" /> or <paramref name="context" /> contain text that cannot be converted to UTF8.
```
Nit - grammar: _contain_ -\> _contains_. (Same comment elsewhere.)

GrabYourPitchforks commented on src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs:
```diff
@@ -0,0 +1,661 @@
+        /// </exception>
+        /// <remarks>
+        ///   <paramref name="label" /> and <paramref name="context" /> will be converted to bytes using the UTF8 encoding.
+        ///   for other encodings, perform the conversion using the desired encoding and use an overload which accept the
```
Nit - grammar: _accept_ -\> _accepts_

GrabYourPitchforks commented on src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationManaged.cs:
```diff
@@ -0,0 +1,118 @@
+
+                if (current <= 0)
+                {
+                    // The key has already been released and zeroed. Do nothing. This may occur when the this
```
Technically this is an observable change: calling `Dispose` multiple times could cause the key to be zeroed while an operation is taking place, which would cause the `DeriveBytes` operation to return an _incorrect_ value (as opposed to throwing an exception). A standard `SafeHandle` implementation works around this by special-casing `Dispose` to only decrement the ref count once.

You could work around it with something like this:

```cs
int _disposed = 0;
public void Dispose()
{
    if (Interlocked.Exchange(ref _disposed, 1) == 0) { ReleaseKey(); }
}
```

It also means that you don't need a loop within `ReleaseKey`. The logic becomes a whole lot easier.

```cs
private void ReleaseKey()
{
    int newCount = Interlocked.Decrement(ref _keyReferenceCount);
    Debug.Assert(newCount >= 0, "How did this go negative?");
    if (newCount == 0) { ZeroKey(); }
}
```

`SafeHandle` has more complexity to keep the "was disposed?" bit _within_ the ref count integer, but that seems like overkill here.

vcsjones replied:
Ohhhhhhhhhhhh good catch. This uncovered a few problems with the reference counting. This along with some asserts makes me feel better about the whole thing.

vcsjones replied:
> but should it be private?

Yes it should have.

GrabYourPitchforks commented:
I've finished review of the prod logic - haven't gotten to tests yet. Some nits, and a race condition in `ReleaseKey` that we should address, but everything LGTM. Will try to get to the tests in a few days.

Thanks Kevin for getting to this! It's awesome to see it come together. :)

> I think there is clear evidence that SP800108 needs to be made available in downlevel targets. What's not clear to me is why everything else couldn't argue that criteria.

Your line of reasoning is valid. I'm seeing some internal requests for AES-GCM (and maybe AES-CCM?) to be made available downlevel as well since folks are clamoring for them on net4x.

vcsjones commented:
@GrabYourPitchforks I believe I have addressed all of your feedback thus far, with most of it actioned, some of it "maybe not". I appreciate your time thus far!

GrabYourPitchforks commented:
> That would probably have to be Windows-only or limited to .NETFramework. Supporting a true netstandard2.0 for GCM / CCM would be a huge multiplier in complexity.

Oh, it would definitely add complexity. But just yesterday I was in an email thread (which I think you were on as well!) where the writing on the wall looms ever larger.

Fortunately that's a problem for future us and not for this PR! 🫠

jeffhandley reviewed (dismissed):
I know we've still been waiting for a re-review from @GrabYourPitchforks, but from what I can tell all previous comments have been addressed. @bartonjs, if this looks good to you, I propose we go ahead and merge this ahead of Preview 1. We can revisit the packaging topic if another idea emerges there, and if there's post-merge feedback from @GrabYourPitchforks, we can address it in a separate PR.

stephentoub commented:
>  We can revisit the packaging topic if another idea emerges there

If we rename it, won't that mean we'll have another dead-end package on nuget forevermore?  It's not the end of the world, just, not ideal.

jeffhandley commented:
> > We can revisit the packaging topic if another idea emerges there
> 
> If we rename it, won't that mean we'll have another dead-end package on nuget forevermore? It's not the end of the world, just, not ideal.

True. I rationalized it as: It'd only exist with `-preview` version(s) and we could unlist it while keeping it available for package restore).

Is `System.Security.Cryptography.Extensions` still a candidate name, or are there objections to that?

vcsjones commented:
> Is `System.Security.Cryptography.Extensions` still a candidate name, or are there objections to that?

My 2 cents, it's my preferred choice.

jeffhandley reviewed (commented):
Nit-picky: It looks like we always use lower-case `.netstandard.cs` and `.netcoreapp.cs` throughout the repo. Can you rename those files to lowercase the suffixes please?

jeffhandley commented:
> > Is `System.Security.Cryptography.Extensions` still a candidate name, or are there objections to that?
> 
> My 2 cents, it's my preferred choice.

@bartonjs is going to re-review, so we'll give him a chance to weigh in on that again. In anticipation of that being OK, I'm tempted to get a commit ready with that restructuring. Would you like to do that, @vcsjones; or would you like me to? (Either's good with me; I have time today).

vcsjones commented:
> Would you like to do that, @vcsjones

I should be able to get that in over the next few hours.

bartonjs commented:
If we want to propose a new package name, that needs to go back to API Review.

bartonjs commented on src/libraries/System.Security.Cryptography.SP800108/src/System/Security/Cryptography/SP800108HmacCounterKdf.cs:
```diff
@@ -0,0 +1,661 @@
+        private readonly SP800108HmacCounterKdfImplementationBase _implementation;
+
+        /// <summary>
+        ///   Initializes a new instance of <see cref="SP800108HmacCounterKdf" /> using key and HMAC algorithm.
```
There's a bit of cave-man speech going on here.  I think the standard form I've seen is

```suggestion
        ///   Initializes a new instance of <see cref="SP800108HmacCounterKdf" /> using a specified key and HMAC algorithm.
```

(also applies to the next ctor)

vcsjones replied:
SP800-108 good. Cave man speech bad!

bartonjs commented:
Since there's active discussion about renaming it, it should go back to API Review.  If this gets merged in the next couple of hours, it makes it into Preview 1... which we don't want to do if we're going to rename it literally tomorrow.  So, despite having signed off on the code, marked the PR as No Merge.

vcsjones commented:
> it should go back to API Review

Okay. I'll address some nits but hold off on pushing the commit that does the rename until it comes back from API review.

vcsjones committed:
SP800108: Basic API shape and Windows CNG implementation

vcsjones committed:
SP800108: Managed implementation provided

vcsjones committed:
SP800108: Add tests

vcsjones committed:
SP800108: Add Microsoft.Bcl.Cryptography

vcsjones committed:
SP800108: Minor fixes

vcsjones commented:
@ViktorHofer added you to get input on the packaging for `Microsoft.Bcl.Cryptography`. The background is we added a new type called `SP800108HmacCounterKdf`. This is inbox in `System.Security.Cryptography`. For down level targets, that package was introduced to contain an implementation for .NET Framework and .NET Standard.

vcsjones committed:
SP800108: Make Windows 7 work because it's easy

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationCng.cs:
```diff
@@ -0,0 +1,213 @@
+        }
+
+        /// <summary>
+        /// Returns null if the platform ins Windows 10+ and psuedo handles should be used.
```
It probably only wants to be a `//` comment instead of xmldoc... but either way "ins" isn't English :smirk:

```suggestion
        /// Returns null if the platform is Windows 10+ and psuedo handles should be used.
```

bartonjs commented on src/libraries/Common/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationManaged.cs:
```diff
@@ -0,0 +1,112 @@
+            {
+                ZeroKey();
+            }
+
```
Nit: Distracting trailing blank line

```suggestion
```

bartonjs commented on src/libraries/Microsoft.Bcl.Cryptography/ref/Microsoft.Bcl.Cryptography.csproj:
```diff
@@ -0,0 +1,17 @@
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp'">
+    <Compile Include="Microsoft.Bcl.Cryptography.cs" />
+  </ItemGroup>
+<ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">
```
Alignment:

```suggestion
  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">
```

bartonjs commented on src/libraries/Microsoft.Bcl.Cryptography/src/System/Security/Cryptography/SP800108HmacCounterKdfImplementationManaged.cs:
```diff
@@ -0,0 +1,202 @@
+                return;
+            }
+
+            byte[] keyBuffer = new byte[key.Length];
```
I was about to ask why keyBuffer is using gcnew here.  Then realized it's probably because IncrementalHash wants a precisely-sized array input.  Probably worth a comment.

vcsjones replied:
Yep. Added a comment.

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/SP800108HmacCounterKdf.Windows.cs:
```diff
@@ -0,0 +1,89 @@
+{
+    public sealed partial class SP800108HmacCounterKdf : IDisposable
+    {
+        private static bool s_isWindows8OrGreater = OperatingSystem.IsWindowsVersionAtLeast(6, 2);
```
Adding `readonly` let's the Tier-N(>0) JIT treat it as a const.

```suggestion
        private static readonly bool s_isWindows8OrGreater = OperatingSystem.IsWindowsVersionAtLeast(6, 2);
```

vcsjones committed:
SP800108: Code review feedback

vcsjones committed:
SP800108: Remove incorrect platform check

ericstj commented:
@vcsjones would this be interesting to add to https://github.com/dotnet/core/issues/8134?  If so please consider adding today before end of day.  Sorry for the late notice.

vcsjones commented:
> would this be interesting to add to https://github.com/dotnet/core/issues/8134?

Eh, I think it's cool, but I don't think it has broad enough appeal, and demonstrating it is tricky.
